[WIP] Elasticache replication group

willthames opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

Add elasticache_replication_group module

##### ISSUE TYPE
 - New Module Pull Request


##### COMPONENT NAME
elasticache_replication_group

##### ANSIBLE VERSION
```
ansible 2.5.0 (devel c2303c5922) last updated 2017/09/21 19:15:21 (GMT +1000)
  config file = None
  configured module search path = [u'/Users/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/will/src/ansible/lib/ansible
  executable location = /Users/will/src/ansible/bin/ansible
  python version = 2.7.13 (default, Dec 18 2016, 07:03:39) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```

##### ADDITIONAL INFORMATION
Tests are hampered a little by Elasticache's absurd constraint that ReplicationGroupId's must be 20 characters max. The typical cloud resource_prefix is far longer than this. Most early tests run well on my machine, although `state: rebooted` and pending modification tests have not made it all of the way through, but not inside docker. Need a `very_short_resource_prefix` for these purposes

One other issue at the moment is that after adding a second cache cluster, we have to wait before later actions in the test suite are possible. This might have implications for the later failover tests and even the teardown, which won't work unless there's a reasonable wait between adding the new cache cluster node and then deleting it.

wimnat committed:
Elasticache replication group initial work

willthames commented:
cc @wimnat - this is based on your work retaining credit in the commit log and authorship.

willthames commented:
This is not the latest commit. Need to push more recent commit

s-hertel commented on lib/ansible/modules/cloud/amazon/elasticache_replication_group.py:
```diff
@@ -0,0 +1,585 @@
+            module.fail_json("Failover enabled requires at least two nodes")
+
+    if state == "present":
+        rep_group = check_for_rep_group(module, connection)
```
You could move this a couple lines above and remove the two in the elifs below.

s-hertel commented on test/integration/targets/elasticache_replication_group/tasks/main.yml:
```diff
@@ -0,0 +1,289 @@
+        aws_access_key: "{{ aws_access_key|default(omit) }}"
+        aws_secret_key: "{{ aws_secret_key|default(omit) }}"
+        security_token: "{{ security_token|default(omit) }}"
+        profile: "{{ profile|default(omit) }}"
```
Since integration tests can't be run with ansible-test with a ~/.aws directory, a profile won't be found if one is provided. Is there a way around that?

s-hertel replied:
^ A way around that: run tests with --docker flag.

willthames replied:
I'll update that to move to standard testing approach now.

s-hertel commented on test/integration/targets/elasticache_replication_group/defaults/main.yml:
```diff
@@ -0,0 +1,3 @@
+# max of 20 chars for elasticache_replication_group_id, not much to play with
+# also, just in case any double hyphens get added (also not allowed), make them single hyphens
+#elasticache_replication_group_id: "{{ (resource_prefix|regex_replace('ansible-test(ing)?', 'test') + '-erg')|replace('--', '-') }}"
```
You could perhaps get the identifying digits of the prefix since that's the helpful part for determining which test it came from with "{{ resource_prefix | regex_findall('.{8}$') }}" and add that to 'test'?

s-hertel commented on lib/ansible/modules/cloud/amazon/elasticache_replication_group.py:
```diff
@@ -0,0 +1,585 @@
+            changed = True
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            msg = "Couldn't modify replication group"
+            if e.response['Error']['Code'] == 'InvalidParameterCombination':
```
There might not be a .response attribute.
Using python3:
```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: AttributeError: 'ParamValidationError' object has no attribute 'response'
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_9utcwaz8/ansible_module_elasticache_replication_group.py\", line 496, in modify_rep_group\n    connection.modify_replication_group(**params)\n  File \"/Users/shertel/Workspace/ansible/venv/python3.5.2/lib/python3.5/site-packages/botocore/client.py\", line 312, in _api_call\n    return self._make_api_call(operation_name, kwargs)\n  File \"/Users/shertel/Workspace/ansible/venv/python3.5.2/lib/python3.5/site-packages/botocore/client.py\", line 575, in _make_api_call\n    api_params, operation_model, context=request_context)\n  File \"/Users/shertel/Workspace/ansible/venv/python3.5.2/lib/python3.5/site-packages/botocore/client.py\", line 630, in _convert_to_request_dict\n    api_params, operation_model)\n  File \"/Users/shertel/Workspace/ansible/venv/python3.5.2/lib/python3.5/site-packages/botocore/validate.py\", line 291, in serialize_to_request\n    raise ParamValidationError(report=report.generate_report())\nbotocore.exceptions.ParamValidationError: Parameter validation failed:\nUnknown parameter in input: \"CacheSubnetGroupName\", must be one of: ReplicationGroupId, ReplicationGroupDescription, PrimaryClusterId, SnapshottingClusterId, AutomaticFailoverEnabled, CacheSecurityGroupNames, SecurityGroupIds, PreferredMaintenanceWindow, NotificationTopicArn, CacheParameterGroupName, NotificationTopicStatus, ApplyImmediately, EngineVersion, AutoMinorVersionUpgrade, SnapshotRetentionLimit, SnapshotWindow, CacheNodeType, NodeGroupId\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/tmp/ansible_9utcwaz8/ansible_module_elasticache_replication_group.py\", line 582, in <module>\n    main()\n  File \"/tmp/ansible_9utcwaz8/ansible_module_elasticache_replication_group.py\", line 568, in main\n    modify_rep_group(module, connection, rep_group)\n  File \"/tmp/ansible_9utcwaz8/ansible_module_elasticache_replication_group.py\", line 500, in modify_rep_group\n    if e.response['Error']['Code'] == 'InvalidParameterCombination':\nAttributeError: 'ParamValidationError' object has no attribute 'response'\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 0}
```

willthames replied:
Yes, that test should only be for ClientError, not BotoCoreError, you're quite right (not a python3 thing, I think that's e.message - one of your comments I just read on another issue reminded me)

anryko commented on lib/ansible/modules/cloud/amazon/elasticache_replication_group.py:
```diff
@@ -0,0 +1,585 @@
+
+    for k, v in rep_group_params.items():
+        if module.params.get(k) is not None and rep_group[v] != module.params.get(k):
+            params[v] = module.params.get[k]
```
Bug. Should be `params[v] = module.params.get(k)`

s-hertel commented on lib/ansible/modules/cloud/amazon/elasticache_replication_group.py:
```diff
@@ -0,0 +1,585 @@
+    required: false
+    default: no
+    choices: [ "yes", "no" ]
+
```
You could add boto3/botocore as requirements.

s-hertel commented on lib/ansible/modules/cloud/amazon/elasticache_replication_group.py:
```diff
@@ -0,0 +1,585 @@
+'''
+
+from ansible.module_utils.ec2 import boto3_conn, get_aws_connection_info, ec2_argument_spec
+from ansible.module_utils.ec2 import HAS_BOTO3, camel_dict_to_snake_dict
```
HAS_BOTO3 doesn't need to be imported since AnsibleAWSModule checks it.

dagwieers committed:
Clean up VMware BOTMETA and author/license info (#36873)

* Clean up vmware BOTMETA

This removes everything that's already part of the module docs. So
basically the BOTMETA file only extends the information from the module.

* Updated author information and copyright statement

orthanc committed:
Additions to AWS Module Guidlines from Pull Request Feedback (#36834)

* Update doc of AnsibleAWSModule to remove incorrect connect example.

Current example uses aws_connect method which no longer exists. Replace
this with the client and resource methods that do exist.

Also remove try/except block in connect example as guidelines imply that
there aren't connection errors thrown on connection, just when later
sing the connection.

* Update AWS module guidelines to include the AnsibleAWSModule connection methods.

* Add information on integration testing to the AWS module guidelines.

* Add information on common documentation fragments to AWS module
Guidelines.

* Changes as requested on PR.

* Restructure connection section to start with current best practice
* Explain the use of the CI build groups
* Use YAML Anchors for aws credentials example

* Replace guidance on use of test groups with link to aliases file.

This achieves the goal of explaining why this file is necessary without
introducing overhead of keeping documentation up to date when test
groups change.

Commit:
Fix IAM setup for profiles and add ReplaceRoute policy

trishnaguha committed:
eos_static_route fix (#36903)

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

ganeshrn committed:
Enable eapi integration test for eos modules (#36934)

kbreit committed:
Reversed orser of icmp_type and icmp_code per bug 36378 (#36380)

ryansb committed:
Use Tower v2 API when available, falling back to v1 where necessary (#36663)

* Use Tower v2 API when available, falling back to v1 where necessary

bcoca committed:
updated api example

- added many more comments
- removed very obsolete 1.x refs

Akasurde committed:
VMware: new module : vmware_datastore_maintenancemode (#36436)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

willthames committed:
Remove old boto documentation and promote AnsibleAWSModule (#36763)

There should no longer be a need for boto documentation - it
just adds significant extra clutter to the guidelines.

AnsibleAWSModule greatly improves the codebase to such an extent
that we should mandate it for new code unless there is a
documented good reason.

willthames committed:
[cloud aws_s3] Enforce required parameters for the various aws_s3 modes (#36767)

[cloud aws_s3] Most modes require `object` parameter, and this is easy to
get wrong (e.g. through confusion with the `dest` parameter). As
it's as easy to enforce, let's do that.

dharmabumstead committed:
Copy edits and tweaks

slapula commented on lib/ansible/modules/cloud/amazon/elasticache_replication_group.py:
```diff
@@ -0,0 +1,585 @@
+    params['Engine'] = module.params.get('engine')
+    params['EngineVersion'] = module.params.get('cache_engine_version')
+    params['CacheParameterGroupName'] = module.params.get('cache_parameter_group')
+    params['CacheSubnetGroupName'] = module.params.get('cache_subnet_group_name')
```
This needs to be `CacheSubnetGroupNames`.  And the associated module parameter should be set to a 'list' type.

slapula commented on lib/ansible/modules/cloud/amazon/elasticache_replication_group.py:
```diff
@@ -0,0 +1,585 @@
+            engine=dict(default='redis', choices=['redis']),
+            cache_engine_version=dict(),
+            cache_parameter_group=dict(),
+            cache_subnet_group_name=dict(),
```
As I mentioned above, this should be `cache_subnet_group_names=dict(type='list'),`

slapula commented on lib/ansible/modules/cloud/amazon/elasticache_replication_group.py:
```diff
@@ -0,0 +1,585 @@
+    params['Engine'] = module.params.get('engine')
+    params['EngineVersion'] = module.params.get('cache_engine_version')
+    params['CacheParameterGroupName'] = module.params.get('cache_parameter_group')
+    params['CacheSubnetGroupName'] = module.params.get('cache_subnet_group_name')
```
This needs to be CacheSubnetGroupNames. And the associated module parameter type should be set to 'list'.

willthames replied:
Do you have any docs to back this up? https://boto3.readthedocs.io/en/latest/reference/services/elasticache.html#ElastiCache.Client.create_replication_group says it takes `CacheSubnetGroupName`, which is a string.

I tried taking your approach and it failed the test suite.

slapula replied:
I guess I was seeing things?  You are correct.  It is indeed not plural and takes a string as opposed to list. 😄

slapula commented on lib/ansible/modules/cloud/amazon/elasticache_replication_group.py:
```diff
@@ -0,0 +1,585 @@
+
+    # data structure of returned security groups from describe_cache_clusters is
+    # different to that expected by modify_replication_group
+    if set(module.params['security_group_ids']) != set([sg['SecurityGroupId'] for sg in cache_cluster['SecurityGroups']]):
```
In my testing, I've found that if a cluster doesn't have any security groups assigned to it, this key will not exist.  This `if` statement then returns an error complaining about this `NoneType` not being iterable.  I did something like this to get around it.  This probably isn't the best way to do it but it works for me:

```
    # data structure of returned security groups from describe_cache_clusters is
    # different to that expected by modify_replication_group
    if 'SecurityGroups' in cache_cluster:
        if set(module.params['security_group_ids']) != set([sg['SecurityGroupId'] for sg in cache_cluster['SecurityGroups']]):
            params['SecurityGroupIds'] = module.params['security_group_ids']

    if 'SecurityGroups' not in cache_cluster:
        if module.params['security_group_ids']:
            params['SecurityGroupIds'] = module.params['security_group_ids']
```

willthames replied:
Fixed

slapula commented on lib/ansible/modules/cloud/amazon/elasticache_replication_group.py:
```diff
@@ -0,0 +1,585 @@
+            engine=dict(default='redis', choices=['redis']),
+            cache_engine_version=dict(),
+            cache_parameter_group=dict(),
+            cache_subnet_group_name=dict(),
```
As I mentioned above, this should be cache_subnet_group_names=dict(type='list'),

willthames replied:
Or not :)

slapula reviewed (changes requested):
I'm so glad someone wrote up this module!  In order for me to make this usable however, I had to make the changes I pointed out in this review.  Let me know if you have any questions.  I'll be using my version of this module in production so I'll make sure I update this PR if I run into any further issues.

Akasurde committed:
New module: digital_ocean_tag_facts (#36965)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

Akasurde committed:
New module: digital_ocean_load_balancer_facts (#36966)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

pmarques committed:
Fix Base64 decode TypeError (#36968)

```
TypeError: expected bytes-like object, not str
```

Commit:
fixes cliconf and netconf plugin loaders

This change updates the configuration and plugin loader objects for
cliconf and netconf to pull be able to configure where to find the
plugins.

Commit:
ugh q

Commit:
adds fix for terminal plugins

smithjlrhit committed:
Assume firewall is offline on DBUS_ERROR during init (#36486)

During init, the FirewallClient tries to connect to the DBUS socket.  If it
fails to connect it should be safe to assume firewalld is offline.

Akasurde committed:
New module: digital_ocean_snapshot_facts (#35897)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

Akasurde committed:
VMware: Refactor vmware_target_canonical_facts (#36964)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

velimir committed:
fix python3 compatibility for znode module

kazoo client requires bytes and not string

resmo committed:
cs_zone_facts: implement return facts as ansible returns (#36993)

abirami-n committed:
Fix_regex (#36569)

Qalthos committed:
Move _start_connection() to module_utils/connection and fix Popen() call (#36249)

* Fix Popen() call to use ansible python

* Remove PATH searching, reuse location of ansible instead

Qalthos committed:
ios_static_route idempotence fix (#35912)

* Remove default admin_distance and fix the idempotence thereof

Fixes #33290

* Fix tests and use yaml anchors to shorten tests

* Add test for undefined admin_distance

* Read config from `show run` if `show ip static route` fails

* Restore flags to ios.get_config &  use get_config where appropriate

gundalow committed:
Ability to link to other pages from plugin docs (#37009)

Support relative links

okgolove committed:
[terraform] Support initializing modules and providers at runtime (#36996)

* Fixes #36994

Added function for force initializing modules and providers

Added type for force_init parameter

fridim committed:
ini_file: add allow_no_value param (#24442)

MikeKlebolt committed:
Fixes commit a28eb94, which broke validate_certs = False for python < 2.7.9 (#34887)

msven committed:
ec2_asg: fix #28087 and #35993 (#36679)

Fixes #35993 - Changes to update_size in commit eb4cc31 made it so
the group dict passed into update_size was not modified. As a result,
the 'replace' call does not see an updated min_size like it previously
did and doesn't pause to wait for any new instances to spin up. Instead,
it moves straight into terminating old instances. Fix is to add batch_size
to min_size when calling wait_for_new_inst.

Fixes #28087 - Make replace_all_instances and replace_instances behave
exactly the same by setting replace_instances = current list of instances
when replace_all_instances used. Root cause of issue was that without lc_check
terminate_batch will terminate all instances passed to it and after updating
the asg size we were querying the asg again for the list of instances - so terminate batch
saw the list including new ones just spun up.

When creating new asg with replace_all_instances: yes and lc_check: false
the instances that are initially created are then subsequently replaced.
This change makes it so replace only occurs if the asg already existed.

Add integration tests for #28087 and #35993.

newswangerd committed:
fix NIOS DNS lookup example (#36737)

gundalow committed:
Support <hr/> in module docs (#37030)

abadger committed:
Python3 issues in maven_artifact (#37035)

Fixes #33761

okgolove committed:
Added variables to destroy command (#37028)

ryansb committed:
Route custom waiter (#36922)

This creates a way for us to use boto3's data-driven waiter support to use custom waiters where Boto3 hasn't implemented them yet.

The only waiter implemented so far is for VPC Route Tables to check that they exist, and this replaces some custom retry code.

Commit:
Elasticache replication group improvements

Add tests, and improve module to meet tests

willthames commented:
Currently getting tests failing after cache creation - at that point in the tests, looking up the replication group clearly fails, so it tries to create the replication group, and fails because it already exists - this is in spite of waiters waiting for the replication group to be available before continuing

slapula commented on lib/ansible/modules/cloud/amazon/elasticache_replication_group.py:
```diff
@@ -0,0 +1,586 @@
+    cache_cluster_params = {
+        'cache_engine_version': 'EngineVersion',
+        'cache_parameter_group': 'CacheParameterGroupName',
+        'cache_subnet_group_name': 'CacheSubnetGroupName',
```
Ok, I found another issue and this time I do have documentation to back me up. 😄 

 `CacheSubnetGroupName` does not exist as a modifiable parameter: http://boto3.readthedocs.io/en/latest/reference/services/elasticache.html#ElastiCache.Client.modify_replication_group

gundalow commented:
Rebase needed https://docs.ansible.com/ansible/latest/dev_guide/developing_rebasing.html

willthames commented:
Closing - I have no need for this and don't have the cycles to fix this back up to required standards
